### DISCUSSION TITLE
HA non si connette a internet se connesso alla VPN

### DISCUSSION THREAD

**Discussion by TruantRiccardo** [indomus/forum#1686] [HUB: Home Assistant]
Ciao a tutti,
ho un raspberry con HA e vivo in una casa con rete FWA (quindi nattata). per poter accedere a HA da remoto ho seguito una guida che ho trovato su inDomus la quale mi ha fatto installare una VPN su un server remoto alla quale mi potevo poi collegare dall'esterno. Tutto perfetto se non che HA non è più in grado di connettersi a internet. Mi han consigliato di aprire delle porte ma ho provato varie combinazioni con nessun risultato( ho provato la 443 in out, la 80 in out la 8123 in out e la porte della vpn in out), ma forse sbaglio in qualche impostazione (e.g. tag di destinazione). C'è qualcuno che mi potrebbe aiutare o che mi potrebbe indicare una strada?
Altra cosa che ho notato è che posso collegarmi al dominio solo tramite la porta 8123, la porta 443 non viene reindirizzata all'IP dove è connesso HA.
Grazie in anticipo,
Riccardo

**Comment by snakuzzo** [comment]
Per quanto riguarda la connessione da HA verso l'esterno, verifichiamo appena possibile, ma credo dipenda dalle impostazioni legate al client openvpn.
Quanto al discorso dele porte 8123 e 443, il reindirizzamento dipende da:

1. l'impostazione delle regole firewall sulla VM su GCP
2. la configurazione del routing con iptables

Tu puoi utilizzare le porte che ti pare, a patto che sia configurato tutto correttamente

**Reply by TruantRiccardo** [reply]
io ho seguito il comando consigliato sulla guida, quindi con iptables, identico per la 8123 che per la 443. 
![image](https://user-images.githubusercontent.com/44238224/167616514-cc90e183-334f-4380-acab-8a89b4b6570b.png)
solo che per la 8123 ha funzionato, per la 443 no. unica differenza che vedo nel firewall è che sono tutte e due aprte in entrata, ma mentre la 8123 ha destinazione my-host, la 443 ha destinazione https-server.
Colgo l'occasione per chiederti chiarimenti sul tag di destinazione: da quello che ho capito server per filtrare la regola a solo le istanze che possiedono quel tag, quindi la 443 entrata la possono sfruttare solo le istanze con tag https-server. Ma avendo la mia VM tag my-host, questo vuol dire che lei non vede la 443 aperta?

**Reply by snakuzzo** [reply]
Tu pensa al giro che farebbe una chiamata verso il tuo FQDN.
La prima cosa che deve attraversare è il firewall della VM su GCP. Lì devi configurare i tag con le porte da inoltrare ed associare quel tag alla VM interessata (o alle VM interessate). A questo punto la chiamata arriva alla VM, ma poi a sua volta deve inoltrare il traffico verso l'IP del tuo server. Ed ecco che entra in ballo iptables con il prerouting.
Più chiaro di così non so spiegartelo

**Reply by snakuzzo** [reply]
Fatto la verifica che ti avevo promesso...

Questa la configurazione di rete sul RPI (oscurato solo il mac address delle scheda di rete)

```
pi@raspberrypi:~/gcp$ ifconfig -a
docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255
        ether 02:42:09:32:19:9d  txqueuelen 0  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether yy:yy:yy:yy:yy:yy  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 3201  bytes 5022379 (4.7 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 3201  bytes 5022379 (4.7 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

tun0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1500
        inet 10.8.0.2  netmask 255.255.255.0  destination 10.8.0.2
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 2627  bytes 194977 (190.4 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4781  bytes 4790272 (4.5 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlan0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.48  netmask 255.255.255.0  broadcast 192.168.1.255
        ether xx:xx:xx:xx:xx:xx  txqueuelen 1000  (Ethernet)
        RX packets 416762  bytes 564932851 (538.7 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 153696  bytes 34158612 (32.5 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

ping verso il server VPN

```
pi@raspberrypi:~/gcp$ ping 10.8.0.1
PING 10.8.0.1 (10.8.0.1) 56(84) bytes of data.
64 bytes from 10.8.0.1: icmp_seq=1 ttl=64 time=134 ms
64 bytes from 10.8.0.1: icmp_seq=2 ttl=64 time=139 ms
^C
--- 10.8.0.1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 134.337/136.537/138.738/2.200 ms
```

Prova di connettività verso l'esterno dall'interno del container di home assistant

```
pi@raspberrypi:~/gcp$ docker exec  -it homeassistant sh
/config # ping www.google.com
PING www.google.com (142.250.184.68): 56 data bytes
64 bytes from 142.250.184.68: seq=0 ttl=115 time=17.936 ms
64 bytes from 142.250.184.68: seq=1 ttl=115 time=20.862 ms
64 bytes from 142.250.184.68: seq=2 ttl=115 time=20.766 ms
^C
--- www.google.com ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 17.936/19.854/20.862 ms
/config # 
```

**Reply by TruantRiccardo** [reply]
Grazie mille per il supporto. Ho provato a rieseguire i test e ho notato un'alta perdita di pacchetti (e un ping più alto).

Questo è il mio test verso la VM:
![image](https://user-images.githubusercontent.com/44238224/168691059-79d4581f-47df-412d-bbcc-1a602788c944.png)


Questo è il mio test verso l'esterno:
![image](https://user-images.githubusercontent.com/44238224/168690833-5630efad-4c9d-4b23-b5f2-d24b3de888eb.png)


In effetti, usando Spotify, notavo che se avviavo la riproduzione dal telefono ci voleva una buona quantità di secondi prima che HA si accorgesse dell'avvio della riproduzione(30-40s). Invece se il comando partiva da HA, la riproduzione partiva immediatamente (direi quindi che il problema sia in RX).

Al momento ho installato NGINX sulla VM (è stato così che ho riacquisito la capacità di parlare con l'esterno). Dall'esterno parlo dalla 443 (o dalla 80 che reindirizza alla 443) verso NGINX, il quale poi riporta sulla 8123 verso HA. 443 e 80 sono aperte sulla VM.

Questa invece è la mia configurazione di rete sul RPI:
![image](https://user-images.githubusercontent.com/44238224/168691928-659dd172-a3fa-4156-93c5-29c9e6451543.png)
dalla quale noto dei pacchetti droppati su eth0.

Hai idea cosa potrebbe causa tutte queste perdite di pacchetti?
So che ti sto dando poche info per aiutarmi a diagnosticare il problema. Se c'è qualcosa che ti potrebbe essere d'aiuto dimmi pure.
Grazie ancora

**Reply by snakuzzo** [reply]
Non ne ho idea

**Reply by Corrado-possieri** [reply]
Ciao Riccardo, potresti per favore spiegare come hai fatto a riacquistare la capacità di parlare con l’esterno tramite NGINX? Sto avendo lo stesso problema ma non ho trovato un modo semplice per aggirarlo.

**Reply by TruantRiccardo** [reply]
Ciao Corrado, dammi qualche giorno e ti scrivo come ho fatto io, al momento non mi ricordo con precisione. Premetto che non ho ben capito come ci sia riuscito e che la connessione non è perfetta per quanto riguarda il ping

**Comment by TruantRiccardo** [comment]
Okay, chiaro, ho modificato i tag puntandoli a my-host.
Per quanto riguardo il forwarding di 443:
![image](https://user-images.githubusercontent.com/44238224/167630563-7ab8407b-b2cd-4425-8782-b8eb8233a306.png)
le richieste sulla 443 vengono inoltrate alla 443 del rasp, cosi come scritto sulla guida.
Però nella guida non abbiamo mai detto al rasp di ascoltare la 443 (presumo in ottica di installarci poi NGINX o caddy), quindi il fatto che, contrariamente quanto scritto sulla guida, io non ricevo ancora risposta alla 443 dovrebbe aver senso.
Correggimi se sbaglio, sono molto un principiante per quanto riguarda i protocolli di rete.
Provando a collegarmi al dominio attraverso la porta 443 con NGINX avviato ricevo questo:
![image](https://user-images.githubusercontent.com/44238224/167628454-100e8675-e21a-412d-92bf-efc8f4ee213f.png)
l'errore non so se sia dovuto ad una errata configurazione di NGINX o all'impossibilità di connessioni in OUT.

Per quanto riguarda l'impossibilità di connettersi ad internet invece non trovo niente di errato sulla configurazione della VM quindi presumo anche io possa essere a livello della VPN, senza però avere idea di come poterti aiutare

**Reply by snakuzzo** [reply]
Te lo dice...hai fatto una chiamata in HTTP verso un servizio esposto in HTTPS

**Reply by TruantRiccardo** [reply]
Scusami ma le mie conoscenze non sono buone per capire cosa volesse dire.
Comunque il problema penso fosse nei certificati: NGNIX su HA andava a leggere i certificati generati dal duckdns di HA(di una precedente configurazione e dominio), non quello sulla VM.

Comunque il problema maggioritario rimane l'assenza di connessione, il fatto della porta 443 l'ho tirato fuori inizialmente solo perchè pensavo potesse essere legato. Per questo problema avresti dei consigli?
Grazie per il supporto comunque

**Reply by snakuzzo** [reply]
Per l'assenza di connessione da HA verso internet ti ho già detto...va verificato con mano, non è roba da suggerimento al volo.
Quando posso, non so ancora quando, faccio delle prove e verifico...ma quando ho tempo.

P.S. questo non è un lavoro ;)

**Reply by TruantRiccardo** [reply]
Ma figurati, scusami se sono stato garbato e anzi, ti ringrazio molto per la tua disponibilità gratuita.
Era più se c'era un indicazione per aiutarmi a darti una mano in ciò.
Comunque sono riuscito a collegarlo, anche se non ho capito perchè.
Ho installato NGINX sulla VM, cancellato la regola di forwarding della 443, e creata all'interno di NGINX. 
Ora se faccio un controllo dei certificati con ssllabs mi dice tutto okay. se provo a collegarmi alla 443 o alla 80 ottengo bad request(forse l'errore è l'IP da inserire in trusted proxies, ma non sono sicuro), però ora sono connesso a internet(meteo, spotify e aggiornamenti funzionano).
Sarei comunque curioso di capire perchè ora va e cosa sbaglio sulla porta 443 :)

**Reply by TruantRiccardo** [reply]
ti confermo che inserendo l'IP locale della VM sono ora in grado di accedere dalla 443. Se vuoi maggiori dettagli su cos'ho fatto con NGINX scrivimi pure, perchè sarei curioso di sapere il motivo :)

**Reply by snakuzzo** [reply]
Cioè? Scrivi pure qui... non ho capito.

**Reply by Corrado-possieri** [reply]
Cerco di essere più preciso. Ho seguito la guida per configurare una VPN a cui connettere HA, ma collegandola non riesco a connettermi ad internet senza utilizzare i reverse proxy. In pratica sembra che il traffico sulla porta 443 sia bloccato.

**Reply by snakuzzo** [reply]
> ti confermo che inserendo l'IP locale della VM sono ora in grado di accedere dalla 443. Se vuoi maggiori dettagli su cos'ho fatto con NGINX scrivimi pure, perchè sarei curioso di sapere il motivo :)

Comunque non capisco cosa c'entri NGINX con la connessione verso internet. NGINX, Caddy o altro di base ti proxano il traffico verso l'interno

**Reply by TruantRiccardo** [reply]
Ciao Snak, scusami ma sono passati un po' di mesi e dovrei rivedere i miei appunti che mi ero lasciato. Quello che ti posso assicurare è che facendo quell'operazione sono passato dall'assenza di connessione internet ad una connessione internet(anche se devo dire che continuo a subire packet loss in entrata ma non in uscita e.g. se do il comando di "pausa spotify" da HA la pausa avviene subito ma HA riceve che la pausa è avvenuta dopo diversi secondi). Martedì ho la consegna della tesi, prossima settimana ti scrivo un breve report su cos'ho fatto e cos'ho ottenuto. Così magari riesco a correggere il problema dei packet loss :)

**Reply by snakuzzo** [reply]
Ciao.
Giusto per rifare il punto...tu sei su HA OS, giusto?
E dopo esserti collegato in VPN, il server dove risiede HA non esce su internet. Corretto?

Chiedo perché anche un altro utente in questi giorni ha lo stesso problema.

**Reply by Corrado-possieri** [reply]
Grazie mille, si sono su HA OS e dopo aver installato la VPN non mi si connette più a internet.

**Reply by snakuzzo** [reply]
Ok. Ma non dicevo a te. Chiedevo info a @TruantRiccardo per capire cos'ha fatto per far funzionare il tutto.
Questo problema pare ci sia solo su HA OS

**Reply by snakuzzo** [reply]
@TruantRiccardo comunque ho appena fatto una prova sulla mia installazione di HA OS (è la prima volta che la provo) e non riscontro problemi di raggiungibilità dal container di core ne dà altri container presenti su HA OS.

Di seguito il log dell'addon...

```
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
Wait until the user uploads the files.
All files available!
Setup the VPN connection with the following OpenVPN configuration.
2022-11-20 14:27:53 DEPRECATED OPTION: --cipher set to 'AES-256-CBC' but missing in --data-ciphers (AES-256-GCM:AES-128-GCM). Future OpenVPN version will ignore --cipher for cipher negotiations. Add 'AES-256-CBC' to --data-ciphers or change --cipher 'AES-256-CBC' to --data-ciphers-fallback 'AES-256-CBC' to silence this warning.
2022-11-20 14:27:53 WARNING: file '/share/client.pwd' is group or others accessible
2022-11-20 14:27:53 OpenVPN 2.5.6 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Apr 17 2022
2022-11-20 14:27:53 library versions: OpenSSL 1.1.1s  1 Nov 2022, LZO 2.10
2022-11-20 14:27:53 Outgoing Control Channel Encryption: Cipher 'AES-256-CTR' initialized with 256 bit key
2022-11-20 14:27:53 Outgoing Control Channel Encryption: Using 256 bit message hash 'SHA256' for HMAC authentication
2022-11-20 14:27:53 Incoming Control Channel Encryption: Cipher 'AES-256-CTR' initialized with 256 bit key
2022-11-20 14:27:53 Incoming Control Channel Encryption: Using 256 bit message hash 'SHA256' for HMAC authentication
2022-11-20 14:27:53 TCP/UDP: Preserving recently used remote address: [AF_INET]xxx.xxx.xxx.xxx:1194
2022-11-20 14:27:53 Socket Buffers: R=[212992->212992] S=[212992->212992]
2022-11-20 14:27:53 UDP link local: (not bound)
2022-11-20 14:27:53 UDP link remote: [AF_INET]xxx.xxx.xxx.xxx:1194
2022-11-20 14:27:56 TLS: Initial packet from [AF_INET]xxx.xxx.xxx.xxx:1194, sid=3d4aaed7 24a16bde
2022-11-20 14:27:56 VERIFY OK: depth=1, CN=ChangeMe
2022-11-20 14:27:56 VERIFY KU OK
2022-11-20 14:27:56 Validating certificate extended key usage
2022-11-20 14:27:56 ++ Certificate has EKU (str) TLS Web Server Authentication, expects TLS Web Server Authentication
2022-11-20 14:27:56 VERIFY EKU OK
2022-11-20 14:27:56 VERIFY X509NAME OK: CN=ha-cloud-vm_5fdcbbaf-5355-4a01-8092-262c03635b1e
2022-11-20 14:27:56 VERIFY OK: depth=0, CN=ha-cloud-vm_5fdcbbaf-5355-4a01-8092-262c03635b1e
2022-11-20 14:27:57 Control Channel: TLSv1.3, cipher TLSv1.3 TLS_AES_256_GCM_SHA384, peer certificate: 256 bit EC, curve prime256v1, signature: ecdsa-with-SHA256
2022-11-20 14:27:57 [ha-cloud-vm_5fdcbbaf-5355-4a01-8092-262c03635b1e] Peer Connection Initiated with [AF_INET]xxx.xxx.xxx.xxx:1194
2022-11-20 14:27:57 PUSH: Received control message: 'PUSH_REPLY,route-gateway 10.8.0.1,topology subnet,ping 15,ping-restart 120,ifconfig 10.8.0.2 255.255.255.0,peer-id 1,cipher AES-256-GCM'
2022-11-20 14:27:57 Pushed option removed by filter: 'route-gateway 10.8.0.1'
2022-11-20 14:27:57 OPTIONS IMPORT: timers and/or timeouts modified
2022-11-20 14:27:57 OPTIONS IMPORT: --ifconfig/up options modified
2022-11-20 14:27:57 OPTIONS IMPORT: peer-id set
2022-11-20 14:27:57 OPTIONS IMPORT: adjusting link_mtu to 1624
2022-11-20 14:27:57 OPTIONS IMPORT: data channel crypto options modified
2022-11-20 14:27:57 Data Channel: using negotiated cipher 'AES-256-GCM'
2022-11-20 14:27:57 Outgoing Data Channel: Cipher 'AES-256-GCM' initialized with 256 bit key
2022-11-20 14:27:57 Incoming Data Channel: Cipher 'AES-256-GCM' initialized with 256 bit key
2022-11-20 14:27:57 TUN/TAP device tun0 opened
2022-11-20 14:27:57 /sbin/ip link set dev tun0 up mtu 1500
2022-11-20 14:27:57 /sbin/ip link set dev tun0 up
2022-11-20 14:27:57 /sbin/ip addr add dev tun0 10.8.0.2/24
2022-11-20 14:27:57 Initialization Sequence Completed
```

Qui il test...

```

| |  | |                          /\           (_)   | |            | |  
| |__| | ___  _ __ ___   ___     /  \   ___ ___ _ ___| |_ __ _ _ __ | |_ 
|  __  |/ _ \| '_ \ _ \ / _ \   / /\ \ / __/ __| / __| __/ _\ | '_ \| __|
| |  | | (_) | | | | | |  __/  / ____ \\__ \__ \ \__ \ || (_| | | | | |_ 
|_|  |_|\___/|_| |_| |_|\___| /_/    \_\___/___/_|___/\__\__,_|_| |_|\__|

Welcome to the Home Assistant command line.

System information
  IPv4 addresses for enp0s3: 192.168.1.9/24
  IPv6 addresses for enp0s3: fe80::1810:ab77:5181:5094/64

  OS Version:               Home Assistant OS 9.3
  Home Assistant Core:      2022.11.3

  Home Assistant URL:       http://homeassistant.local:8123
  Observer URL:             http://homeassistant.local:4357
➜  ~ 
➜  ~ 
➜  ~ nc -zvw2 google.com 443 
google.com (142.250.184.78:443) open
➜  ~ docker exec -it homeassistant bash 
bash-5.1# nc -zvw2 google.com 443
google.com (142.250.184.78:443) open
```

Piccola nota, per lo start dell'addon date un'occhiata a questa issue -> https://github.com/Ailme/homeassistant-openvpn-client-addon/issues/6#issuecomment-1225852234
Bisogna aggiungere l'addon in local facendo un clone del progetto e cambiare un parametro nel file di configurazione `config.json`, altrimenti l'addon non parte. C'è una PR già creata per risolvere il problema, ma purtroppo il mantainer del progetto non si fa vivo da anni...quindi per ora fate come descritto a quel link se avete questo problema.

Se può esservi utile vi aggiungo il mio file di configurazione per la connessione al server vpn (al netto dei certificati)...

```
client
dev tun
proto udp
remote miodominio.duckdns.org 1194
resolv-retry infinite
pull-filter ignore "route-gateway"
nobind
remote-cert-tls server
tls-version-min 1.2
verify-x509-name ha-cloud-vm_zxczxczxczxczxczxczxc name
cipher AES-256-CBC
auth SHA256
auth-nocache
askpass /share/client.pwd
ping-restart 10
verb 3
```

...ed il mio file di configurazione sulla VM su GCP (sempre al netto dei certificati)...

```
dev tun
proto udp
port 1194
ca /etc/openvpn/easy-rsa/pki/ca.crt
cert /etc/openvpn/easy-rsa/pki/issued/ha-cloud-vm_zxczxczxczxczxczxczxc.crt
key /etc/openvpn/easy-rsa/pki/private/ha-cloud-vm_zxczxczxczxczxczxczxc.key
dh none
ecdh-curve prime256v1
topology subnet
server 10.8.0.0 255.255.255.0
# Set your primary domain name server address for clients
#push "dhcp-option DNS 9.9.9.9"
#push "dhcp-option DNS 149.112.112.112"
# Prevent DNS leaks on Windows
#push "block-outside-dns"
# Override the Client default gateway by using 0.0.0.0/1 and
# 128.0.0.0/1 rather than 0.0.0.0/0. This has the benefit of
# overriding but not wiping out the original default gateway.
#push "redirect-gateway def1"
client-to-client
client-config-dir /etc/openvpn/ccd
keepalive 15 120
remote-cert-tls client
tls-version-min 1.2
tls-crypt /etc/openvpn/easy-rsa/pki/ta.key
cipher AES-256-CBC
auth SHA256
user openvpn
group openvpn
persist-key
persist-tun
crl-verify /etc/openvpn/crl.pem
status /var/log/openvpn-status.log 20
status-version 3
syslog
verb 3
#DuplicateCNs allow access control on a less-granular, per user basis.
#Remove # if you will manage access by user instead of device. 
#duplicate-cn
# Generated for use by PiVPN.io
```

**NOTA IMPORTANTE:** nella configurazione del server ricordatevi di commentare le seguenti righe...

```
# Set your primary domain name server address for clients
#push "dhcp-option DNS 9.9.9.9"
#push "dhcp-option DNS 149.112.112.112"
# Prevent DNS leaks on Windows
#push "block-outside-dns"
# Override the Client default gateway by using 0.0.0.0/1 and
# 128.0.0.0/1 rather than 0.0.0.0/0. This has the benefit of
# overriding but not wiping out the original default gateway.
#push "redirect-gateway def1"
```

Quelle push in sostanza mandano "quelle" configurazioni al client che si connette al server.
Ma, nel nostro caso specifico, il client non deve usare il default gw della VPN per uscire su internet, ma deve usare quello suo di default.
Stesso dicasi per l'utilizzo dei DNS, deve usare quelli suoi e non quelli che vuole assegnargli il server VPN
